### PR TITLE
Pass sessionTrackingEnabled with all non-Analytics API requests

### PR DESCRIPTION
**Pull request by tmeyer2115**
This CR ensures that the sessionTrackingEnabled param is passed along with all non-Analytics API requests issued by the SDK. The ApiRequest class now expects sessionTrackingEnabled as part of the configuration object passed to its constructor. If this value is not present in the config, an error will be thrown. This offers a bit of future proofing by ensuring that anyone adding new
ApiRequests must pass in a sessionTrackingEnabled param. This should resolve issue #419 .

J=ANS-436
TEST=auto, manual

Updated Jest tests. Verified that I saw the sessionTrackingEnabled param with the following requests:
    - Vertical Autocomplete
    - Universal Autocomplete
    - Vertical Search
    - Universal Search
    - Error reporting
    - Q&A submission

### CODE REVIEW

**Comment by data-enabler** [comment]
> I removed the setSessionsOptIn method from answers-umd. Users can already set the sessionTrackingEnabled variable in their Answers config and adding the ability to dynamically change this value didn't seem especially valuable.

Was this discussed with Product? At any rate, we shouldn't be changing the public api in a bugfix release.

**Comment by tmeyer2115** [comment]
@data-enabler Added the setSessionsOptIn method back. Passed globalStorage to the various services so that they can pass an up-to-date sessionTrackingEnabled value when constructing ApiRequests.

File: src/answers-umd.js [hunk 1]
```diff
@@ -123,81 +123,79 @@ class Answers {
   }
 
   init (config) {
-    config = Object.assign({}, DEFAULTS, config);
-    if (typeof config.apiKey !== 'string') {
+    const parsedConfig = this.parseConfig(config);
+    if (typeof parsedConfig.apiKey !== 'string') {
```
Comment (apokellypse) [inline src/answers-umd.js]: would recommend the error checking of the config be factored out into a validateConfig method, similar to how you did parseConfig, which was nice.

**Review by apokellypse — DISMISSED** [review]

**Review by data-enabler — COMMENTED** [review]
Also, prefixing the commit message with "Answers SDK" is pretty redundant.

File: src/core/http/apirequest.js [hunk 2]
```diff
@@ -24,6 +25,16 @@ export default class ApiRequest {
...
 
+    /**
+     * A boolean indicating if sessionTracking is enabled
+     * @type {boolean}
+     * @private
+     */
+    if (typeof opts.sessionTrackingEnabled !== 'boolean') {
+      throw new AnswersBasicError('Must indicate if session tracking is enabled', 'ApiRequest');
+    }
+    this._sessionTrackingEnabled = opts.sessionTrackingEnabled;
```
Comment (data-enabler) [inline src/core/http/apirequest.js]: When I was talking about passing globalStorage through, I meant passing it all the way through to ApiRequest. It's not a very clean solution either way, but at least we won't have boilerplate all over the place.

Also, we should leave a TODO somewhere in this class to make it clear that this isn't really a desired state, and we want to move to something else like a service interface with dependency injection.

File: src/answers-umd.js [hunk 3]
```diff
@@ -226,6 +224,21 @@ class Answers {
...
 
+  /**
+   * Parses the config provided by the user. In the parsed config, any options not supplied by the
+   * user are given default values.
+   * @param {Object} config The user supplied config.
+   */
+  parseConfig (config) {
+    const parsedConfig = Object.assign({}, DEFAULTS, config);
+    let sessionTrackingEnabled = true;
+    if (typeof config.sessionTrackingEnabled === 'boolean') {
```
Comment (data-enabler) [inline src/answers-umd.js:229]: This is fine for now, but In the long term we should probably be validating the config with JSON Schema or something, so we'll be able to use Object.assign to handle the defaults for this as well.

**Review by data-enabler — CHANGES_REQUESTED** [review]

File: src/answers-umd.js (hunk 1 above)
Comment (tmeyer2115) [reply src/answers-umd.js]: Done.

File: src/core/http/apirequest.js (hunk 2 above)
Comment (tmeyer2115) [reply src/core/http/apirequest.js]: Done.

File: src/core/http/apirequest.js [hunk 4]
```diff
@@ -62,6 +53,18 @@ export default class ApiRequest {
...
      */
     this._params = opts.params || {};
+
+    // TOOD (tmeyer): Create an interface for fetching sitesTrackingEnabled. Inject an
+    // implementation of that interface here.
+    /**
+     * The global storage of the Answers experience
+     * @type {GlobalStorage}
+     * @private
+     */
```
Comment (data-enabler) [inline src/core/http/apirequest.js:65]: JSDoc needs to go immediately before the declaration. I would also get rid of the superfluous description.

File: src/core/http/apirequest.js [hunk 5]
```diff
@@ -62,6 +53,18 @@ export default class ApiRequest {
      * @private
      */
     this._params = opts.params || {};
+
+    // TOOD (tmeyer): Create an interface for fetching sitesTrackingEnabled. Inject an
+    // implementation of that interface here.
```
Comment (data-enabler) [inline src/core/http/apirequest.js]: Not really what I meant. An `ApiService` implementation should get passed as a dependency to `ErrorReporter`, `AutoCompleteApi`, etc. It's fine for the `ApiService` implementation to have a reference to `GlobalStorage`.

**Review by data-enabler — CHANGES_REQUESTED** [review]

**Review by data-enabler — APPROVED** [review]

**Pull request merged**